Add support for batch requests to JSON RPC server

RaoulSchaffranek opened the pull request:
Simbolik and kontrol-node exchange many requests, and processing them one request at a time quickly becomes a performance bottleneck. By batching requests together we can avoid the bottleneck. The batching implementation here is part of the JSON RPC spec, see: https://www.jsonrpc.org/specification#batch

RaoulSchaffranek committed:
Add support for batch requests to JSON RPC server

RaoulSchaffranek committed:
Run pyupgrade

RaoulSchaffranek committed:
Make format and make check

tothtamas28 commented on pyk/src/pyk/rpc/rpc.py:
```diff
@@ -71,37 +72,105 @@ class JsonRpcMethod(Protocol):
-        super().__init__(*args, **kwargs)
+    method: str
+    params: Any
+    id: Any
```
```suggestion
    id: int | str
```

tothtamas28 replied:
Also consider making `id` the first attribute.

tothtamas28 commented on pyk/src/pyk/rpc/rpc.py:
```diff
@@ -71,37 +72,105 @@ class JsonRpcMethod(Protocol):
-class JsonRpcRequestHandler(BaseHTTPRequestHandler):
-    methods: dict[str, JsonRpcMethod]
+@dataclass(frozen=True)
+class JsonRpcRequest:
```
Consider just making this a `typing.NamedTuple`:

```suggestion
class JsonRpcRequest(NamedTuple):
```

tothtamas28 commented on pyk/src/tests/unit/test_json_rpc.py:
I think this should rather be an integration test.

tothtamas28 commented on pyk/src/pyk/rpc/rpc.py:
```diff
@@ -71,37 +72,105 @@ class JsonRpcMethod(Protocol):
+    id: Any
+
+    @staticmethod
+    def validate(request_dict: Any, valid_methods: Iterable[str]) -> JsonRpcRequest | JsonRpcError:
```
Consider moving this to `JsonRpcRequestHandler`.

tothtamas28 replied:
Then `valid_methods` can also be removed from the parameters (assuming the method is not static).

tothtamas28 commented on pyk/src/pyk/rpc/rpc.py:
```diff
@@ -113,44 +182,44 @@ def do_POST(self) -> None:  # noqa: N802
+        if isinstance(validation_result, JsonRpcError):
+            return validation_result
 
+        method_name = request['method']
```
```suggestion
        method_name = validation_result.method
```

tothtamas28 replied:
But if `JsonRpcRequest` is a `NamedTuple`, you could also just unpack:

```python3
method, args, id = JsonRpcRequest.validate(request, self.methods.keys())
```

RaoulSchaffranek committed:
Address code review comments

RaoulSchaffranek commented:
Thanks for the review @tothtamas28! Sounds all reasonable to me and I addressed your comments in https://github.com/runtimeverification/k/pull/4688/commits/f8fa54f823fe9202e2a86efd8deb40d77413b872

tothtamas28 commented on pyk/src/pyk/rpc/rpc.py:
```diff
@@ -71,37 +72,80 @@ class JsonRpcMethod(Protocol):
+class JsonRpcResult:
+
+    def encode(self) -> bytes:
+        raise NotImplementedError('Subclasses must implement this method')
```
```suggestion
class JsonRpcResult(ABC):

    @abstractmethod
    def encode(self) -> bytes: ...
```

RaoulSchaffranek committed:
Annotate abstractmethod
